ui/next: Metrics Query Reducer

mrtracy opened the pull request:
Add reducer to handle metrics query data into UI/next. Includes extensive tests.
- A "MetricsQuery" represents the current metrics data needed by a single
  component (e.g. a graph). The metrics reducer maintains a collection of these,
  organized by ID.
- Components can request new metrics by dispatching the "queryMetrics"
  asynchronous action. This method is designed to batch requests with the same
  timespan before fetching from the server.

Also in this commit:
- Add time series protos to generated protobuffer output.
- Configure `typings.json` to install our generated protobuffer typings into our
  normal set of typings.
- Add a `local_typings` folder, which contains a modified version of the typings
  for long.d.ts. The typings file provided by DefinitelyTyped appears to be
  broken; we should attempt to upstream this fix later.
- Begin using "isomorphic fetch" instead of "whatwg-fetch".
- Begin using "fetch-mock" library to mock server responses in our mocha tests.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/6511)

<!-- Reviewable:end -->

Commit:
ui/next: Metrics Query Reducer

Add reducer to handle metrics query data into UI/next. Includes extensive tests.

+ A "MetricsQuery" represents the current metrics data needed by a single
component (e.g. a graph). The metrics reducer maintains a collection of these,
organized by ID.
+ Components can request new metrics by dispatching the "queryMetrics"
asynchronous action. This method is designed to batch requests with the same
timespan before fetching from the server.

Also in this commit:

+ Add time series protos to generated protobuffer output.
+ Configure `typings.json` to install our generated protobuffer typings into our
normal set of typings.
+ Add a `local_typings` folder, which contains a modified version of the typings
for long.d.ts. The typings file provided by DefinitelyTyped appears to be
  broken; we should attempt to upstream this fix later.
+ Begin using "isomorphic fetch" instead of "whatwg-fetch".
+ Begin using "fetch-mock" library to mock server responses in our mocha tests.

tamird commented:
infra stuff LG

---

Reviewed 11 of 13 files at r1.
Review status: 11 of 13 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[ui/next/typings.json, line 5 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511#-KGxiHKjh10-GpPUxf-B:-KGxiHKjh10-GpPUxf-C:-460668065) ([raw file](https://github.com/cockroachdb/cockroach/blob/b26da5845fea7cea58020986d1c2f9c8966f809d/ui/next/typings.json#L5)):*
neat.

---

*[ui/next/typings.json, line 8 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511#-KGxiF8uIFsPaFQf82-s:-KGxiF8uIFsPaFQf82-t:-1983687851) ([raw file](https://github.com/cockroachdb/cockroach/blob/b26da5845fea7cea58020986d1c2f9c8966f809d/ui/next/typings.json#L8)):*
put a todo here

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511#-:-KGxitvM0UPOyEbB_QVz:-936157252)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 3 of 3 files at r2.
Review status: 11 of 13 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 13 files at r1.
Review status: all files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---

*[ui/next/app/redux/metrics.spec.ts, line 152 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511#-KGy1N7-etFHyqG5JDPB:-KGy1N7-etFHyqG5JDPC:1156010393) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca09f1ca16aab5df4c3f502c80606027f70ee98b/ui/next/app/redux/metrics.spec.ts#L152)):*
ditto this any - isn't it known to be a [`response`](https://developer.mozilla.org/en-US/docs/Web/API/Response)?

---

*[ui/next/app/redux/metrics.ts, line 8 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511#-KGxzAz7AWqyrF0FoBd-:-KGxzAz8t7XD2ZQlLvDC:-1820429294) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca09f1ca16aab5df4c3f502c80606027f70ee98b/ui/next/app/redux/metrics.ts#L8)):*
how come this is `require`d? i believe this is `import`ed elsewhere. is it just because the other location is untested?

---

*[ui/next/app/redux/metrics.ts, line 85 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511#-KGxzmMzN6yKSDFWKwpG:-KGxzmMzN6yKSDFWKwpH:-602809484) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca09f1ca16aab5df4c3f502c80606027f70ee98b/ui/next/app/redux/metrics.ts#L85)):*
can you hoist this down out of the switch?

---

*[ui/next/app/redux/metrics.ts, line 114 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511#-KGxzqf-XQ3nJAGEB4cw:-KGxzqf-XQ3nJAGEB4cx:-469765787) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca09f1ca16aab5df4c3f502c80606027f70ee98b/ui/next/app/redux/metrics.ts#L114)):*
ditto

---

*[ui/next/app/redux/metrics.ts, line 152 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511#-KGxzt4MEFb423tpWapw:-KGxzt4MEFb423tpWapx:-469765787) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca09f1ca16aab5df4c3f502c80606027f70ee98b/ui/next/app/redux/metrics.ts#L152)):*
ditto

---

*[ui/next/app/redux/metrics.ts, line 281 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511#-KGy--N9WYt3_a6St8bz:-KGy--N9WYt3_a6St8c-:-1246552833) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca09f1ca16aab5df4c3f502c80606027f70ee98b/ui/next/app/redux/metrics.ts#L281)):*
why the any?

---

*[ui/next/app/redux/metrics.ts, line 282 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511#-KGy-BKUL7MCrpk7R5UQ:-KGy-BKUL7MCrpk7R5UR:136486228) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca09f1ca16aab5df4c3f502c80606027f70ee98b/ui/next/app/redux/metrics.ts#L282)):*
isn't response typed? specifically, isn't it a [`response`](https://developer.mozilla.org/en-US/docs/Web/API/Response)?

---

*[ui/next/app/redux/metrics.ts, line 284 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511#-KGy-60n4JaEoCve2W_f:-KGy-60n4JaEoCve2W_g:1855008172) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca09f1ca16aab5df4c3f502c80606027f70ee98b/ui/next/app/redux/metrics.ts#L284)):*
I think you can use the TimeSeriesQueryResponse interface here instead of any

---

*[ui/next/app/redux/metrics.ts, line 317 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511#-KGy-XGbaP7Vk42y7I-T:-KGy-XGbaP7Vk42y7I-U:1692855131) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca09f1ca16aab5df4c3f502c80606027f70ee98b/ui/next/app/redux/metrics.ts#L317)):*
do you really need this?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511)*
<!-- Sent from Reviewable.io -->

maxlang commented:
Isomorphic fetch sounded like it depended on whatwg-fetch for the polyfill, if I'm looking at the right one (https://github.com/matthew-andrews/isomorphic-fetch)

Other than that <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> with some minor comments

---

Review status: all files reviewed at latest revision, 26 unresolved discussions, some commit checks failed.

---

*[ui/next/app/redux/metrics.spec.ts, line 49 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511#-KH-4i2_O0eewrAehHgu:-KH-4i2_O0eewrAehHgv:1950567109) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca09f1ca16aab5df4c3f502c80606027f70ee98b/ui/next/app/redux/metrics.spec.ts#L49)):*
Would it be better to assign queries the value `{}`?

---

*[ui/next/app/redux/metrics.spec.ts, line 192 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511#-KH-7-rTNP7EWfUdSNyL:-KH-7-rTNP7EWfUdSNyM:1757531480) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca09f1ca16aab5df4c3f502c80606027f70ee98b/ui/next/app/redux/metrics.spec.ts#L192)):*
Why do we want `currentRequest` defined here?

---

*[ui/next/app/redux/metrics.spec.ts, line 262 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511#-KH-7lRkxb6HS38mRfnf:-KH-7lRkxb6HS38mRfng:964285725) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca09f1ca16aab5df4c3f502c80606027f70ee98b/ui/next/app/redux/metrics.spec.ts#L262)):*
I might add one test for what happens if there are two separate calls at different ticks. For example:

```
p1 = queryMetrics("id.1", createRequest(timespan, "key.1"));
setTimeout(() => {
  p2 = queryMetrics("id.1", createRequest(timespan, "key.1"));
  ...
})

---

*[ui/next/app/redux/metrics.ts, line 15 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511#-KGzudglohG1bj4cyckz:-KGzudglohG1bj4cycl-:1123520523) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca09f1ca16aab5df4c3f502c80606027f70ee98b/ui/next/app/redux/metrics.ts#L15)):*
Not urgent, but naming these `Request` and `Response` is confusing me a little bit because it's the same as ES6's `Request` and `Response`.

---

*[ui/next/app/redux/metrics.ts, line 22 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511#-KGzmvsVqTvGQ6XkRdj2:-KGzmvsVqTvGQ6XkRdj3:-800133475) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca09f1ca16aab5df4c3f502c80606027f70ee98b/ui/next/app/redux/metrics.ts#L22)):*
Maybe rename this to `FETCH_BEGIN` or `FETCH_ONGOING` to distinguish it from `REQUEST` which actually performs the `fetch`?

---

*[ui/next/app/redux/metrics.ts, line 29 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511#-KGzzzjPZXc3ULqq01Ok:-KGzzzjPZXc3ULqq01Ol:-1532572896) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca09f1ca16aab5df4c3f502c80606027f70ee98b/ui/next/app/redux/metrics.ts#L29)):*
Idle thought about typescript best practices/patterns.

What if this was just a simple interface:

```
interface HasID {
  id: string;
}
```

Then you could recreate `WithID<Request>` as `HasID & Request`. The main benefit is that you get to avoid nesting the data, so the id lives at the same level. I think this neatens up some of the code.

Not saying you should change it, just curious what you thought.

---

*[ui/next/app/redux/metrics.ts, line 41 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511#-KGzodyU8shTHCCp17C6:-KGzodyU8shTHCCp17C7:-1521882478) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca09f1ca16aab5df4c3f502c80606027f70ee98b/ui/next/app/redux/metrics.ts#L41)):*
Could probably drop the "current" from every property here.

---

*[ui/next/app/redux/metrics.ts, line 83 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511#-KGzpQfPlG136KtAKo0e:-KGzpQfQQyNkZrr9FDop:-1861506205) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca09f1ca16aab5df4c3f502c80606027f70ee98b/ui/next/app/redux/metrics.ts#L83)):*
Do we want to clear out `currentRequest` after `RECEIVE` or `ERROR`?

---

*[ui/next/app/redux/metrics.ts, line 101 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511#-KGzve8VVYm6DahaT3Gx:-KGzve8VVYm6DahaT3Gy:-176113073) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca09f1ca16aab5df4c3f502c80606027f70ee98b/ui/next/app/redux/metrics.ts#L101)):*
I feel like I'm making lots of naming nitpicks. `metricQueriesReducer` vs `metricsQueryReducer` is a little confusing. Maybe `allMetricQueriesReducer` and `singleMetricsQueryReducer`.

---

*[ui/next/app/redux/metrics.ts, line 108 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511#-KGzqeS-3xm6Fyk7Oomy:-KGzqeS-3xm6Fyk7Oomz:-1513909665) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca09f1ca16aab5df4c3f502c80606027f70ee98b/ui/next/app/redux/metrics.ts#L108)):*
Would `<WithId<any>>` be more precise here?

---

*[ui/next/app/redux/metrics.ts, line 134 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511#-KGzsNXlAyRwxeWnoJWV:-KGzsNXmikHE_GrmhwUG:374883182) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca09f1ca16aab5df4c3f502c80606027f70ee98b/ui/next/app/redux/metrics.ts#L134)):*
Would it make sense to pull out the inFlight status to a separate reducer and combine them with `combineReducers`?

```
combineReducers({
  inFlight: inFlightReducer,
  queries: metricQueriesReducer,
})
```

---

*[ui/next/app/redux/metrics.ts, line 275 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511#-KGzy7c8dUVCa-EYDr7p:-KGzy7c8dUVCa-EYDr7q:-2083190124) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca09f1ca16aab5df4c3f502c80606027f70ee98b/ui/next/app/redux/metrics.ts#L275)):*
It feels strange to clone when all you really want is start/end_nanos.

---

*[ui/next/app/redux/metrics.ts, line 277 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511#-KGzwcSC4bQzHFBJp0q4:-KGzwcSC4bQzHFBJp0q5:309188315) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca09f1ca16aab5df4c3f502c80606027f70ee98b/ui/next/app/redux/metrics.ts#L277)):*
Does `_.flatMap(batch, (req) => req.data.queries)` work?

One thing I miss from the non-typescript world was that the `_.map(batch, "data.queries")` syntax ever slightly more concise, even though it's so minor.

I've used that syntax sometimes with casts, but then when people rename object properties, for example, the type checking doesn't work and you lose the benefit of typescript.

---

*[ui/next/app/redux/metrics.ts, line 284 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511#-KGy-60n4JaEoCve2W_f:-KH-0z4WVawnRzUzfKnX:1525060777) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca09f1ca16aab5df4c3f502c80606027f70ee98b/ui/next/app/redux/metrics.ts#L284)):*
Might need to use `response.json<TimeSeriesQueryResponse>()` on the line above for it to work?

---

*[ui/next/app/redux/metrics.ts, line 300 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511#-KH-3PikFgxlvgGyWqOu:-KH-3JS1On_r4K4cpBCL:1136199372) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca09f1ca16aab5df4c3f502c80606027f70ee98b/ui/next/app/redux/metrics.ts#L300)):*
You can just use `results.splice(0, numQueries)` here and remove the drop below.

---

*[ui/next/app/redux/nodes.ts, line 39 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511#-KGznRvoTuyiGAOnY0Vi:-KGznRvoTuyiGAOnY0Vj:-1448017582) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca09f1ca16aab5df4c3f502c80606027f70ee98b/ui/next/app/redux/nodes.ts#L39)):*
Would

```function reducer(state: NodeStatusState = new NodeStatusState(), ...)```

work?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511#-:-KGzv4seDAesB-sEnSzx:-1636456687)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: 9 of 17 files reviewed at latest revision, 26 unresolved discussions, some commit checks pending.

---

*[ui/next/typings.json, line 8 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511#-KGxiF8uIFsPaFQf82-s:-KH0tbVNTc5vfLPcQTXU:-1520168808) ([raw file](https://github.com/cockroachdb/cockroach/blob/b26da5845fea7cea58020986d1c2f9c8966f809d/ui/next/typings.json#L8)):*
Can you even put a comment in this file? It doesn't like '//' or '/**/'

---

*[ui/next/app/redux/metrics.spec.ts, line 49 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511#-KH-4i2_O0eewrAehHgu:-KH0cEhAyZONcUQqlIgs:-2129091624) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca09f1ca16aab5df4c3f502c80606027f70ee98b/ui/next/app/redux/metrics.spec.ts#L49)):*
metricQueriesReducer has a default argument of `new metricQueriesReducer()`

---

*[ui/next/app/redux/metrics.spec.ts, line 152 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511#-KGy1N7-etFHyqG5JDPB:-KH0u0DAfVrhUgmCqkZ_:-831583517) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca09f1ca16aab5df4c3f502c80606027f70ee98b/ui/next/app/redux/metrics.spec.ts#L152)):*
The type cannot be inferred by the call to fetchMock.mock alone. Putting a type here is essentially the samy as casting it.

What's coming here is not a `response`, but a `RequestInit`, which is defined by isomorphic-fetch (this is mocking the call to fetch, not the http system).  We can cast to RequestInit, but the 'body' property of RequestInit is `Blob | FormData | string`, and JSON.parse only accepts strings, meaning we have to change the call below to `JSON.parse(request.body as string)`.  Therefore, we would still be casting every usage of `requestObj` or any of its properties: it's easier to leave it as an explicit `any` in this case.

---

*[ui/next/app/redux/metrics.spec.ts, line 192 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511#-KH-7-rTNP7EWfUdSNyL:-KH0cL9hu-kpkWC4J0uM:-2010207108) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca09f1ca16aab5df4c3f502c80606027f70ee98b/ui/next/app/redux/metrics.spec.ts#L192)):*
The updated request gets stored immediately (not the results). This allows components to know that they've already made a request, if they get re-rendered before the fetch is complete.

---

*[ui/next/app/redux/metrics.ts, line 8 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511#-KGxzAz7AWqyrF0FoBd-:-KH0ekPI7LsoEFT5u1kD:-1416287564) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca09f1ca16aab5df4c3f502c80606027f70ee98b/ui/next/app/redux/metrics.ts#L8)):*
This is the more correct notation, it should be updated in the other places, I just forgot to. Updated everywhere to be `require` for this lib.

---

*[ui/next/app/redux/metrics.ts, line 22 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511#-KGzmvsVqTvGQ6XkRdj2:-KH0evi4F84HoJWcKPlE:437652267) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca09f1ca16aab5df4c3f502c80606027f70ee98b/ui/next/app/redux/metrics.ts#L22)):*
REQUEST doesn't perform a fetch, the `queryMetrics()` action creator does.  This is really only used internally (exported for testing).

---

*[ui/next/app/redux/metrics.ts, line 29 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511#-KGzzzjPZXc3ULqq01Ok:-KH0fCc5n4Pv9mgR3Ey-:1452238717) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca09f1ca16aab5df4c3f502c80606027f70ee98b/ui/next/app/redux/metrics.ts#L29)):*
The ID only exists in the scope of this file; the data the 'WithID' contains comes from somewhere else, sometimes over the wire. We would have to mutate that structure as it comes into this file (adding an id), or create a new structure when sending it somewhere else (removing the id).  I don't think this makes the code neater.

---

*[ui/next/app/redux/metrics.ts, line 41 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511#-KGzodyU8shTHCCp17C6:-KH16v2zBHR73Ut3upO8:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca09f1ca16aab5df4c3f502c80606027f70ee98b/ui/next/app/redux/metrics.ts#L41)):*
Done.

---

*[ui/next/app/redux/metrics.ts, line 83 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511#-KGzpQfPlG136KtAKo0e:-KH0fw2wNYxAJjgBLeEg:-482821188) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca09f1ca16aab5df4c3f502c80606027f70ee98b/ui/next/app/redux/metrics.ts#L83)):*
No, 'currentRequest' is the primary way to determine if a component needs to request new data.

---

*[ui/next/app/redux/metrics.ts, line 85 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511#-KGxzmMzN6yKSDFWKwpG:-KH0gECcVjeMtwAEOi9p:-30093231) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca09f1ca16aab5df4c3f502c80606027f70ee98b/ui/next/app/redux/metrics.ts#L85)):*
This is idiomatic in redux; the linter also complains about not having a default case.

---

*[ui/next/app/redux/metrics.ts, line 101 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511#-KGzve8VVYm6DahaT3Gx:-KH16wvWxyXZ_r2qQ1-M:-263369832) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca09f1ca16aab5df4c3f502c80606027f70ee98b/ui/next/app/redux/metrics.ts#L101)):*
Changed the collection names to `MetricQuerySet` and `metricQuerySetReducer`.

---

*[ui/next/app/redux/metrics.ts, line 108 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511#-KGzqeS-3xm6Fyk7Oomy:-KH0zl3nuRo5tAzPlxRR:1887670405) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca09f1ca16aab5df4c3f502c80606027f70ee98b/ui/next/app/redux/metrics.ts#L108)):*
I considered that, I'll go ahead and use it.

---

*[ui/next/app/redux/metrics.ts, line 114 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511#-KGxzqf-XQ3nJAGEB4cw:-KH0h0cgJ-JR5WSrZeNW:-30093231) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca09f1ca16aab5df4c3f502c80606027f70ee98b/ui/next/app/redux/metrics.ts#L114)):*
This is idiomatic in redux; the linter also complains about not having a default case.

---

*[ui/next/app/redux/metrics.ts, line 134 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511#-KGzsNXlAyRwxeWnoJWV:-KH0h2jgg4-EZphege_t:-989817783) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca09f1ca16aab5df4c3f502c80606027f70ee98b/ui/next/app/redux/metrics.ts#L134)):*
Possibly, but let's worry about that when we actually make a component which displays things based on the inFlight states.

---

*[ui/next/app/redux/metrics.ts, line 152 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511#-KGxzt4MEFb423tpWapw:-KH0x0KD3HR48qbxKc4w:-1582418045) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca09f1ca16aab5df4c3f502c80606027f70ee98b/ui/next/app/redux/metrics.ts#L152)):*
See above.

---

*[ui/next/app/redux/metrics.ts, line 275 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511#-KGzy7c8dUVCa-EYDr7p:-KH0x4LoBirU3Wf4gCXf:691775596) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca09f1ca16aab5df4c3f502c80606027f70ee98b/ui/next/app/redux/metrics.ts#L275)):*
Clone also keeps the object prototype (it's not just a pojo, it's a class from the protojs library).

---

*[ui/next/app/redux/metrics.ts, line 277 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511#-KGzwcSC4bQzHFBJp0q4:-KH0xD89ya5VX8fgTAbq:-424298085) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca09f1ca16aab5df4c3f502c80606027f70ee98b/ui/next/app/redux/metrics.ts#L277)):*
flatMap loses typing information (and that may be fundamental to lodash), I don't like using it when type-safe alternatives are available.

---

*[ui/next/app/redux/metrics.ts, line 281 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511#-KGy--N9WYt3_a6St8bz:-KH0hDRqENRJU96FcsC1:1809024565) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca09f1ca16aab5df4c3f502c80606027f70ee98b/ui/next/app/redux/metrics.ts#L281)):*
Did your patch get accepted to the typescript generator? Apparently it did and we have access to the `.encodeJSON()` now. Fixed.

---

*[ui/next/app/redux/metrics.ts, line 282 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511#-KGy-BKUL7MCrpk7R5UQ:-KH0xdwKpOd4HDXvvgyZ:1307907577) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca09f1ca16aab5df4c3f502c80606027f70ee98b/ui/next/app/redux/metrics.ts#L282)):*
Yes, it's an IResponse from isomorphic fetch, and it is inferred correctly.  Fixed.

---

*[ui/next/app/redux/metrics.ts, line 284 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511#-KGy-60n4JaEoCve2W_f:-KH0yAjqHVhSLr-xTQuH:-248408296) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca09f1ca16aab5df4c3f502c80606027f70ee98b/ui/next/app/redux/metrics.ts#L284)):*
We don't use the value directly, we immediately throw it into a `new TimeSeriesQueryResponse(json)`, which already does the cast from `any` for us.  There is no benefit to casting it earlier.

---

*[ui/next/app/redux/metrics.ts, line 300 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511#-KH-3PikFgxlvgGyWqOu:-KH0zTkUhk1HDfjsL5xD:2101108240) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca09f1ca16aab5df4c3f502c80606027f70ee98b/ui/next/app/redux/metrics.ts#L300)):*
Yes, that seems to work better. Fixed.

---

*[ui/next/app/redux/metrics.ts, line 317 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511#-KGy-XGbaP7Vk42y7I-T:-KH0hXaMJ4YFzHpzma8v:1915836059) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca09f1ca16aab5df4c3f502c80606027f70ee98b/ui/next/app/redux/metrics.ts#L317)):*
Yes, we want the inFlight counter so that we (at some point) give some subtle indication that requests are happening.  Also, in this particular case I need it in order to synchronize my tests - the idiomatic redux example on testing thunks uses the same paattern.

---

*[ui/next/app/redux/nodes.ts, line 39 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511#-KGznRvoTuyiGAOnY0Vi:-KH0zfL45eiSsV4NJfaW:-871306678) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca09f1ca16aab5df4c3f502c80606027f70ee98b/ui/next/app/redux/nodes.ts#L39)):*
Yes, I forgot to switch this when I adopted the new pattern. Fixed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 8 of 8 files at r3.
Review status: all files reviewed at latest revision, 24 unresolved discussions, some commit checks failed.

---

*[ui/next/typings.json, line 8 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511#-KGxiF8uIFsPaFQf82-s:-KH1AWPX6tOvddV0h-1R:-1728916910) ([raw file](https://github.com/cockroachdb/cockroach/blob/b26da5845fea7cea58020986d1c2f9c8966f809d/ui/next/typings.json#L8)):*
Bummer. Can you put the TODO in log.d.ts, then?

---

*[ui/next/app/redux/metrics.spec.ts, line 152 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511#-KGy1N7-etFHyqG5JDPB:-KH1ABZOFrxm87nIh7_T:-1153536540) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca09f1ca16aab5df4c3f502c80606027f70ee98b/ui/next/app/redux/metrics.spec.ts#L152)):*
Dunno, the alternative strategy you lay out seems better to me (since it lets future developers leverage code completion when writing mocks).

---

*[ui/next/app/redux/metrics.ts, line 85 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511#-KGxzmMzN6yKSDFWKwpG:-KH1B33poL0MfVzA6a9I:1420666516) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca09f1ca16aab5df4c3f502c80606027f70ee98b/ui/next/app/redux/metrics.ts#L85)):*
what's idiomatic, the default case or the duplication of the `return` call? seems like you could just leave the default clause empty.

---

*[ui/next/app/redux/metrics.ts, line 114 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511#-KGxzqf-XQ3nJAGEB4cw:-KH1BC8TqITEtF83xD8Y:-1284216453) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca09f1ca16aab5df4c3f502c80606027f70ee98b/ui/next/app/redux/metrics.ts#L114)):*
same question

---

*[ui/next/app/redux/metrics.ts, line 152 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511#-KGxzt4MEFb423tpWapw:-KH1BOabqFZ7SK8IGdTC:-1284216453) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca09f1ca16aab5df4c3f502c80606027f70ee98b/ui/next/app/redux/metrics.ts#L152)):*
same question

---

*[ui/next/app/redux/metrics.ts, line 277 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511#-KGzwcSC4bQzHFBJp0q4:-KH1CAKpz-30NtzqSpmS:-205206803) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca09f1ca16aab5df4c3f502c80606027f70ee98b/ui/next/app/redux/metrics.ts#L277)):*
not that flatMap is probably more efficient.

---

*[ui/next/app/redux/metrics.ts, line 284 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511#-KGy-60n4JaEoCve2W_f:-KH1CJu5UjO-Bc7D2e5O:181403118) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca09f1ca16aab5df4c3f502c80606027f70ee98b/ui/next/app/redux/metrics.ts#L284)):*
the benefit is that it's more precise - it documents the shape of that object.

---

*[ui/next/app/redux/metrics.ts, line 317 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511#-KGy-XGbaP7Vk42y7I-T:-KH1CaM9plTL6bv5zuHn:-1207518320) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca09f1ca16aab5df4c3f502c80606027f70ee98b/ui/next/app/redux/metrics.ts#L317)):*
not sure what you think I meant, but I was only asking about the explicit `return`

---

*[ui/next/app/redux/metrics.ts, line 279 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511#-KH1Bdvy39uSEzrsd3wH:-KH1Bdvy39uSEzrsd3wI:-1776850511) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c4d952ca5dddfb9c3ead245d006845f7f6ccb10/ui/next/app/redux/metrics.ts#L279)):*
is this `new` because TSRequest is cockroach.ts.TimeSeriesQueryRequest and not protos.cockroach.ts.TimeSeriesQueryRequest?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: all files reviewed at latest revision, 22 unresolved discussions, some commit checks failed.

---

*[ui/next/typings.json, line 8 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511#-KGxiF8uIFsPaFQf82-s:-KH1HihNzwrmo72mF_Pk:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/b26da5845fea7cea58020986d1c2f9c8966f809d/ui/next/typings.json#L8)):*
Done.

---

*[ui/next/app/redux/metrics.spec.ts, line 152 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511#-KGy1N7-etFHyqG5JDPB:-KH1HWx47hjf7cyNB-Qr:-497477713) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca09f1ca16aab5df4c3f502c80606027f70ee98b/ui/next/app/redux/metrics.spec.ts#L152)):*
Fine.

---

*[ui/next/app/redux/metrics.spec.ts, line 262 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511#-KH-7lRkxb6HS38mRfnf:-KH1FaYO9clHFGpRHzeV:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca09f1ca16aab5df4c3f502c80606027f70ee98b/ui/next/app/redux/metrics.spec.ts#L262)):*
Done.

---

*[ui/next/app/redux/metrics.ts, line 15 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511#-KGzudglohG1bj4cyckz:-KH1FbOXjB7LBTKDrdxH:-1865004519) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca09f1ca16aab5df4c3f502c80606027f70ee98b/ui/next/app/redux/metrics.ts#L15)):*
Renamed these to "TSRequest" and "TSResponse". Wasn't even using either "Query" or "Result", removed them.

---

*[ui/next/app/redux/metrics.ts, line 85 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511#-KGxzmMzN6yKSDFWKwpG:-KH1FjH5C2LZe3M5Rnjp:1509338141) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca09f1ca16aab5df4c3f502c80606027f70ee98b/ui/next/app/redux/metrics.ts#L85)):*
The default case with a return call is idiomatic.

---

*[ui/next/app/redux/metrics.ts, line 277 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511#-KGzwcSC4bQzHFBJp0q4:-KH1FyiiAKAsrCtXgf4p:-1925459765) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca09f1ca16aab5df4c3f502c80606027f70ee98b/ui/next/app/redux/metrics.ts#L277)):*
"probably" doesn't mean "definitely", this is plenty efficient and I get type safety.

---

*[ui/next/app/redux/metrics.ts, line 284 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511#-KGy-60n4JaEoCve2W_f:-KH1GKXueBA9eJE6HAhN:-497477713) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca09f1ca16aab5df4c3f502c80606027f70ee98b/ui/next/app/redux/metrics.ts#L284)):*
Fine.

---

*[ui/next/app/redux/metrics.ts, line 317 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511#-KGy-XGbaP7Vk42y7I-T:-KH1GaGudlZMq2ycVTF6:-60481972) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca09f1ca16aab5df4c3f502c80606027f70ee98b/ui/next/app/redux/metrics.ts#L317)):*
Thought you were talking about the entire promise thing.  No, the return isn't necessary, removed.

---

*[ui/next/app/redux/metrics.ts, line 279 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511#-KH1Bdvy39uSEzrsd3wH:-KH1GGBfl-cUDeZffN2I:677251276) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c4d952ca5dddfb9c3ead245d006845f7f6ccb10/ui/next/app/redux/metrics.ts#L279)):*
Yes.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 3 of 3 files at r4.
Review status: all files reviewed at latest revision, 18 unresolved discussions, some commit checks pending.

---

*[ui/next/app/redux/metrics.ts, line 277 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511#-KGzwcSC4bQzHFBJp0q4:-KH1KOrfKVXqajK_AQcJ:1779507336) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca09f1ca16aab5df4c3f502c80606027f70ee98b/ui/next/app/redux/metrics.ts#L277)):*
Fine, but it is definitely more efficient.

---

*[ui/next/app/redux/metrics.ts, line 279 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511#-KH1Bdvy39uSEzrsd3wH:-KH1K4t3jprQBwlWD7ew:-1950225428) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c4d952ca5dddfb9c3ead245d006845f7f6ccb10/ui/next/app/redux/metrics.ts#L279)):*
What I meant to say was that these interfaces (and all internal interfaces AFAICT) should operate on `protos.cockroach.ts.TimeSeriesQueryRequest` because that's the object's specific type, while `cockroach.ts.TimeSeriesQueryRequest`is an interface that seems more useful as a cast target for incoming JSON objects from the wire.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: 15 of 17 files reviewed at latest revision, 18 unresolved discussions, some commit checks pending.

---

*[ui/next/app/redux/metrics.ts, line 279 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511#-KH1Bdvy39uSEzrsd3wH:-KH1MfZtQ4FU9a3nX_S2:-215131871) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c4d952ca5dddfb9c3ead245d006845f7f6ccb10/ui/next/app/redux/metrics.ts#L279)):*
I have replaced basically every usage of `TSRequest/TSResponse` with `TSRequestMessage/TSResponseMessage`, except in the case where a JSON object is being received off of the wire.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511)*
<!-- Sent from Reviewable.io -->

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> mod the one spurious `new`

---

Reviewed 2 of 2 files at r5.
Review status: all files reviewed at latest revision, 18 unresolved discussions, some commit checks pending.

---

*[ui/next/app/redux/metrics.ts, line 279 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511#-KH1Bdvy39uSEzrsd3wH:-KH1NLHycFT3z4ng1mb4:-799282072) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c4d952ca5dddfb9c3ead245d006845f7f6ccb10/ui/next/app/redux/metrics.ts#L279)):*
thanks. so this can be `unifiedRequest.encodeJSON()` now?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511#-:-KH1NgiqMkUFgO2J0CYK:2113017364)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: 16 of 17 files reviewed at latest revision, 18 unresolved discussions, some commit checks pending.

---

*[ui/next/app/redux/metrics.ts, line 279 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511#-KH1Bdvy39uSEzrsd3wH:-KH1PH14KFn6NrEmNFj3:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c4d952ca5dddfb9c3ead245d006845f7f6ccb10/ui/next/app/redux/metrics.ts#L279)):*
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r6.
Review status: 16 of 17 files reviewed at latest revision, 17 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511)*
<!-- Sent from Reviewable.io -->

maxlang commented:
Review status: 15 of 17 files reviewed at latest revision, 17 unresolved discussions, some commit checks pending.

---

*[ui/next/app/redux/metrics.spec.ts, line 49 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511#-KH-4i2_O0eewrAehHgu:-KH1WZpc0YLZKq1vOOvn:1289099933) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca09f1ca16aab5df4c3f502c80606027f70ee98b/ui/next/app/redux/metrics.spec.ts#L49)):*
Fair enough, though right now the default state is just `{}`.

---

*[ui/next/app/redux/metrics.ts, line 83 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511#-KGzpQfPlG136KtAKo0e:-KH1Xz_08Gev99bxZGO4:-2140209940) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca09f1ca16aab5df4c3f502c80606027f70ee98b/ui/next/app/redux/metrics.ts#L83)):*
So how does this data get invalidated?

---

*[ui/next/app/redux/metrics.ts, line 277 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511#-KGzwcSC4bQzHFBJp0q4:-KH1YtftWIgIjwyEdswc:1596186311) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca09f1ca16aab5df4c3f502c80606027f70ee98b/ui/next/app/redux/metrics.ts#L277)):*
This surprises me somehow.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6511)*
<!-- Sent from Reviewable.io -->
